Sources/WalletConnect/Engine/SessionEngine.swift

```diff
@@ -282,46 +246,37 @@ final class SessionEngine {
+        proposerToRequestPayload[proposal.proposer.publicKey] = payload
+        onSessionProposal?(proposal)
+    }
+        
+    func respondSessionPropose(proposal: SessionType.ProposeParams) {
+        guard let payload = proposerToRequestPayload[proposal.proposer.publicKey] else {
             return
         }
         
-        let settledTopic: String
-        let agreementKeys: AgreementSecret
+        let selfPublicKey = try! kms.createX25519KeyPair()
+        var agreementKey: AgreementSecret!
+        
         do {
-            let publicKey = try session.getPublicKey()
-            agreementKeys = try kms.performKeyAgreement(selfPublicKey: publicKey, peerPublicKey: approveParams.responder.publicKey)
-            settledTopic = agreementKeys.derivedTopic()
-            try kms.setAgreementSecret(agreementKeys, topic: settledTopic)
+            agreementKey = try kms.performKeyAgreement(selfPublicKey: selfPublicKey, peerPublicKey: proposal.proposer.publicKey)
         } catch {
             relayer.respondError(for: payload, reason: .missingOrInvalid("agreement keys"))
             return
         }
 
-        let proposal = pendingSession.proposal
-        let settledSession = SessionSequence.buildAcknowledged(approval: approveParams, proposal: proposal, agreementKeys: agreementKeys, metadata: metadata)
-        sequencesStore.delete(topic: proposal.topic)
-        sequencesStore.setSequence(settledSession)
-        wcSubscriber.setSubscription(topic: settledTopic)
-        wcSubscriber.removeSubscription(topic: proposal.topic)
+        let sessionTopic = agreementKey.derivedTopic()
         
-        let approvedSession = Session(
-            topic: settledTopic,
-            peer: approveParams.responder.metadata,
-            permissions: Session.Permissions(
-                methods: pendingSession.proposal.permissions.jsonrpc.methods), accounts: settledSession.settled!.accounts, expiryDate: settledSession.expiryDate, blockchains: settledSession.settled!.blockchain)
+        let pendingSession = SessionSequence.buildResponded(proposal: proposal, agreementKeys: agreementKey, metadata: metadata, topic: sessionTopic)
         
-        logger.debug("Responder Client approved session on topic: \(topic)")
-        relayer.respondSuccess(for: payload)
-        onSessionApproved?(approvedSession)
+        sequencesStore.setSequence(pendingSession)
+
+        try! kms.setAgreementSecret(agreementKey, topic: sessionTopic)
+        wcSubscriber.setSubscription(topic: sessionTopic)
+        let proposeResponse = SessionType.ProposeResponse(relay: proposal.relay, responder: AgreementPeer(publicKey: selfPublicKey.hexRepresentation))
+        let response = JSONRPCResponse<AnyCodable>(id: payload.wcRequest.id, result: AnyCodable(proposeResponse))
+        relayer.respond(topic: sessionTopic, response: .response(response)) { _ in }
```

Comment:
I think you're responding the proposal over the wrong topic. The proposal must be responded on topic A, while the settlement request that is sent on topic B.